Add documentation regarding annotations (#6362)
inmanta/inmanta-core: docs/lsm/attributes_metadata/attributes_metadata.rst

Attributes modifiers can also be specified on :ref:`relational attributes<attribute_modifiers_on_a_relationship>`.

------------

Annotations
~~~~~~~~~~~

Definition
##########

> Annotations are key-value pairs that can be associated with an entity (service entity or embedded entity) or
> an attribute (simple attribute or relational attribute). These annotations are intended for the web-console to
> improve visualizations and the user-experience.

Reviewer: ```suggestion
Annotations are key-value pairs that can be associated with an entity (service entity or embedded entity) or an attribute (simple attribute or relational attribute). These annotations  don't influence the behavior of LSM or the Inmanta Service Orchestrator itself, but are intended to pass meta data to other components. For example, they can be used to pass on visualization meta-data to the the web-console to improve the user-experience.
```

not sure about this, just a suggestion.

Author: I wasn't sure how to write this part, because the use cases are open ended at the moment. I like the suggestion,

Annotations can be set on embedded entities in the same way.

.. code-block:: inmanta

    entity Interface extends lsm::ServiceEntity:
        string interface_name
        dict __annotations = {"annotation": "value"}
    end

Annotations on simple attributes
################################

Annotations can be attached to simple (non-relational) attributes by defining an attribute of type dict, with a name
``<attribute>__annotations``, where ``<attribute>`` is the name of the attribute the annotations belong to. This
attribute needs a default value containing the attributes. The values of the elements in the dictionary must be
strings.

Example
#######

The example below shows how the annotation ``annotation=value`` is set on the attribute ``interface_name``.
Annotations can be set on simple attributes of embedded entities in the same way.

.. code-block:: inmanta

    entity Interface extends lsm::ServiceEntity:
        string interface_name
        dict interface_name__annotations = {"annotation": "value"}
    end

Annotations on relational attributes
####################################

> Annotations can be attach to a relational attribute by replacing the ``--`` part of the relationship definition with

Reviewer 2: ```suggestion
Annotations can be attached to a relational attribute by replacing the ``--`` part of the relationship definition with
```

an instance of the ``lsm::RelationAnnotations`` entity. This entity has a dict attribute ``annotations`` that

> represent the annotation that should be set on the relational attribute. The values of this dictionary must

Reviewer 2: ```suggestion
represents the annotation that should be set on the relational attribute. The values of this dictionary must
```

> be strings. The annotations are always attached to the attribute that is exposed via the lsm API, i.e. the attribute

Reviewer 2: I don't get what you mean by this sentence. Are you trying to say that you can't / it doesn't make sense to annotate non-API attributes/relations?

Reviewer: me neither

Author: I mean: It's a bidirectional relationship, but only the attribute from the parent to the child is part of the API to create/modify/delete the service instance.

Reviewer 2: I'm not entirely sure but I feel like it may sow more confusion than it fixes as it stands now.

Author: ok. I will remove it.
